core/src/cluster_type.rs

```diff
@@ -0,0 +1,60 @@
+use mongodb::bson::{doc, Bson, Document};
+use mongodb::Client;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
+pub enum MongoClusterType {
+    AtlasDataFederation,
+    Community,
+    Enterprise,
+    UnknownTarget,
+}
+
+pub async fn determine_cluster_type(client: &Client) -> MongoClusterType {
+    let db = client.database("admin");
+
+    // The { buildInfo: 1 } command returns information that indicates
+    // the type of the cluster.
+    let build_info_cmd = doc! { "buildInfo": 1 };
+    let cmd_res: Document = match db.run_command(build_info_cmd).await {
+        Ok(res) => res,
+        Err(e) => {
+            log::error!("Failed to run buildInfo command: {:?}", e);
+            return MongoClusterType::UnknownTarget;
+        }
+    };
+
+    // if "ok" is not 1, then the target type could not be determined.
+    match cmd_res.get("ok") {
+        Some(Bson::Double(f)) if *f == 1.0 => {}
+        Some(Bson::Int32(i)) if *i == 1 => {}
```

Comment:
Should we also include a `Bson::Int64` case?

Reply:
Sure I'll add that.